pylearner.ts

```diff
@@ -4776,8 +5755,194 @@ async function testExamples(examples: Example[]) {
+      throw new Error("Test alias violation case failed, caught incorrect error");
+  }
+  if (!exceptionCaught)
+    throw new Error("Test alias violation case failed, expected an error to be thrown");
+}
+
+async function testAliasingViolationExamples() {
+  await testAliasingViolationExample(aliasViolationExampleIfLocalVariable, "Deze opdracht die het list-object b muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als b", 541, 542);
+  await testAliasingViolationExample(aliasViolationExampleDoubleIfs, "Deze opdracht die het list-object c muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als c", 954, 955);
+  await testAliasingViolationExample(aliasViolationExampleWhileWithIf, "Deze opdracht die het list-object d muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als d", 1263, 1264);
+  await testAliasingViolationExample(aliasViolationExampleIfBasic, "Deze opdracht die het list-object a muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als a", 420, 421);
+  await testAliasingViolationExample(aliasViolationExampleLocalVarsInWhile, "Deze opdracht die het list-object a muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als a", 363, 364);
+  await testAliasingViolationExample(aliasViolationExampleViolationInLastLoopOfLus, "Deze opdracht die het list-object a muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als a", 870, 871);
+  await testAliasingViolationExample(aliasViolationExampleDoubleWhileLusMultipleLoopings, "Deze opdracht die het list-object a muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als a", 719, 720);
+  await testAliasingViolationExample(aliasViolationExampleSingleWhileLusMultipleLoopings, "Deze opdracht die het list-object b muteert wordt met deze preconditie niet ondersteund door Bewijssilhouettencontroleur want de preconditie vermeldt een variabele die mogelijks wijst naar hetzelfde object als b", 680, 681);
+  console.log("All alias violation error tests passed!");
+}
+
 declare var secretExamples: Example[]|undefined;
 
+async function runUnitTests() {
+  function assert(b: boolean) {
+    if (!b)
+      throw new Error("Test case assertion failed");
+  }
+
```

Comment:
Ik denk dat je deze test cases veel beknopter en leesbaarder kunt maken door de volgende definities toe te voegen:
```typescript
function R(mayAliasSets) { return new MyAliasRelation(mayAliasSets); }
const S = ImmutableSet.of;
```
Dan zien de eerste vier asserts er zo uit:
```typescript
assert(R(S()).equals(R(S()));
assert(R(S(S("a", "b"))).equals(R(S(S("a", "b")))));
assert(R(S(S("a"))).equals(R(S()));
assert(R(S(S("b", "a", "c"), S("d", "e", "c"))).equals(R(S(S("a", "b", "c"), S("c", "d", "e")))));
```
Dit maakt het m.i. veel makkelijker om te zien wat hier gebeurt.

Reply:
Ik heb dit eens geprobeerd, maar S als const gebruik als hierboven lukte niet. Ik krijg dan foutmelding dat "TypeError: 'this' is not a function" bij het runnen van de tests, bij het formuleren van een R(S()), omdat het verwacht dat S een functie is door de haakjes () denk ik dan. Ik heb het eens geschreven als een functie en nu werkt het wel, met onderstaande code.
 
`function R(mayAliasSets?: any) {return new MayAliasRelation(mayAliasSets); }`
 `function S(...elements: any[]) { return ImmutableSet.of(elements);}`
  `assert(R(S()).equals(R(S())));`
  `assert(R(S(S("a", "b"))).equals(R(S(S("a", "b")))));`
  `assert(R(S(S("a"))).equals(R(S())));`
  `assert(R(S(S("b", "a", "c"), S("d", "e", "c"))).equals(R(S(S("a", "b", "c"), S("c", "d", "e")))));`

Reply:
Deze oplossing klopte niet. Bij deze formulering voor R en S is elke opgemaakte may-alias-relation één met een lege set aan may-alias-sets. De vier asserts gaven een verkeerd beeld, aangezien elke set leeg is. Zo geeft de equals elke keer True terug. Het probleem zat bij de formuleriing van S met het type van '...elements'. Ik heb nog eens geprobeerd om met een constante te werking voor Immutable Set. Het is gelukt voor een constanste van Immutable Set te maken, wel zonder de .of-methode. Bij de asserts wordt dan het volgende gebruikt.
`assert(R(S.of()).equals(R(S.of())));`